src/hooks/useQuestionnaireResponseDraft.ts

```diff
@@ -0,0 +1,281 @@
+        }, 500);
+
+        return () => {
+            debouncedSaveDraftRef.current?.cancel();
+        };
+    }, []);
+
+    const onUpdateDraft = useCallback(async (questionnaireResponse: QuestionnaireResponse) => {
+        if (!isRunningDebouncedSaveDraftRef.current) {
+            debouncedSaveDraftRef.current?.(questionnaireResponse);
+        }
+    }, []);
+
+    const draftQuestionnaireResponseRD: RemoteData<WithId<QuestionnaireResponse> | undefined> = useMemo(() => {
+        if (isSuccess(response)) {
+            const resultQR = {
+                ...questionnaireResponse,
+                ...response.data,
+                questionnaire: questionnaireId,
+                id: draftId,
+            };
+            return success(resultQR);
+        }
+        if (isFailure(response)) {
+            return success(questionnaireResponse);
+        }
+        return loading;
+        // return response;
+    }, [response, questionnaireId, draftId, questionnaireResponse]);
+
+    const deleteDraft = useCallback(async () => {
+        isRunningDebouncedSaveDraftRef.current = true;
+
+        debouncedSaveDraftRef.current?.cancel();
+
+        await deleteQuestionnaireResponseDraft(draftKeyRef.current, qrDraftServiceType);
+
+        setDraftInfoMessage(undefined);
+        await manager.reloadAsync();
+        isRunningDebouncedSaveDraftRef.current = false;
+    }, [manager, qrDraftServiceType]);
+
+    return {
+        saveDraft,
+        deleteDraft,
+        draftQuestionnaireResponseRD,
+        draftInfoMessage,
+        onUpdateDraft,
+    };
+};
+
+export function getQuestionnaireResponseDraftKeyPrefix(props: {
+    subject?: Resource | Reference | string;
+    questionnaireResponseId?: Resource['id'];
+    qrDraftServiceType: QuestionnaireResponseDraftService;
+}) {
+    const { subject, questionnaireResponseId, qrDraftServiceType } = props;
+
+    if (qrDraftServiceType === 'server') {
+        return questionnaireResponseId;
+    }
+
+    if (questionnaireResponseId) {
+        return `QuestionnaireResponse/${questionnaireResponseId}`;
+    }
+
+    const subjectRef = generateReferenceFromResourceReferenceString(subject);
+    return subjectRef?.reference;
+}
+
+export function generateReferenceFromResourceReferenceString(resource?: Resource | Reference | string) {
+    if (!resource) {
+        return undefined;
+    }
+
+    if (typeof resource === 'string') {
+        return { reference: resource };
+    }
+
+    if (isReference(resource)) {
+        return resource;
+    }
+
+    return getReference(resource);
+}
+
+export const loadQuestionnaireResponseDraft = async (
+    id: Resource['id'],
+    qrDraftServiceType: QuestionnaireResponseDraftService,
+) => {
+    const draftLoadService = getQuestionnaireResponseDraftServices(qrDraftServiceType).loadService;
+    return await draftLoadService(id);
+};
+
+export const saveQuestionnaireResponseDraft = async (
+    keyId: Resource['id'],
+    draftId: Resource['id'],
```

Comment:
remove draft id

Reply:
@alexlipovka I re-thought this code, and we really draftId here while saving instead of while fetching, because while fetdching for the first time we dont't have draft, we have qr with empty id that will trigger populate